.pre-commit-config.yaml

```diff
@@ -0,0 +1,22 @@
+repos:
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v2.3.0
+    hooks:
+    -   id: check-yaml
+    -   id: end-of-file-fixer
+    -   id: trailing-whitespace
+-   repo: https://gitlab.com/pycqa/flake8
+    rev: 3.7.8
+    hooks:
+    -   id: flake8
```

Comment:
Right now I have:

```
$ time flake8 --exclude=sklearn/externals/ --select=F401 sklearn/ examples/
sklearn/build/lib.linux-x86_64-3.7/sklearn/externals/joblib/externals/cloudpickle/__init__.py:3:1: F401 '.cloudpickle.*' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/covariance/tests/test_graphical_lasso.py:20:1: F401 'numpy.testing.assert_equal' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/covariance/tests/test_graph_lasso.py:22:1: F401 'numpy.testing.assert_equal' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/cluster/tests/test_affinity_propagation.py:9:1: F401 'sklearn.utils.testing.assert_equal' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/cluster/tests/test_affinity_propagation.py:9:1: F401 'sklearn.utils.testing.assert_false' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/cluster/tests/test_affinity_propagation.py:9:1: F401 'sklearn.utils.testing.assert_true' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/cross_decomposition/tests/test_pls.py:4:1: F401 'sklearn.utils.testing.assert_equal' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/cross_decomposition/tests/test_pls.py:4:1: F401 'sklearn.utils.testing.assert_true' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/compose/_column_transformer.py:19:1: F401 '..pipeline._fit_one_transformer' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/tests/test_calibration.py:10:1: F401 'sklearn.utils.testing.ignore_warnings' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/decomposition/tests/test_kernel_pca.py:5:1: F401 'sklearn.utils.testing.assert_less' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/decomposition/tests/test_kernel_pca.py:5:1: F401 'sklearn.utils.testing.assert_equal' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/decomposition/tests/test_kernel_pca.py:5:1: F401 'sklearn.utils.testing.assert_not_equal' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/decomposition/tests/test_kernel_pca.py:5:1: F401 'sklearn.utils.testing.ignore_warnings' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/decomposition/tests/test_truncated_svd.py:10:1: F401 'sklearn.utils.testing.assert_equal' imported but unused
sklearn/build/lib.linux-x86_64-3.7/sklearn/decomposition/tests/test_truncated_svd.py:10:1: F401 'sklearn.utils.testing.assert_greater' imported but unused
flake8 --exclude=sklearn/externals/ --select=F401 sklearn/ examples/  284.13s user 2.85s system 960% cpu 29.868 total
```

And as you see it takes quite a while. I'm happy to have these commit hooks, but I guess they shouldn't take too long. I'm not sure how long the other ones take.

Reply:
pre-commit hooks won't run flake8 on all the code base, only on the files that you changed. So that should be mostly immediate.

Reply:
Having existing F401 flake8 errors in the code base is fine, but if you edit one of those files it will ask you to fix the corresponding error.